src/main/java/org/rocstreaming/roctoolkit/NativeObject.java

```diff
@@ -1,18 +1,101 @@
 package org.rocstreaming.roctoolkit;
 
-class NativeObject {
+import static java.util.concurrent.TimeUnit.SECONDS;
 
-    private long ptr;
+/**
+ * A <code>NativeObject</code> represents an underlying native roc object.
+ */
+class NativeObject implements AutoCloseable {
 
-    NativeObject() {
-        ptr = 0L;
+    /**
+     * Maximum time to wait (in milliseconds) for joining {@link AutoCloseThread}.
+     */
+    private final static long MAX_JOIN_TIMEOUT_MS = SECONDS.toMillis(20L);
+
+    /**
+     * <code>NativeObject</code> finalizer thread.
+     */
+    private final static AutoCloseThread thread = AutoCloseThread.getInstance();
+
+    /**
+     *  Underlying roc object native pointer.
+     */
+    private final long ptr;
+
+    /**
+     *  Destructor method.
+     */
+    private final Destructor destructor;
+
+    /**
+     *  Reference to {@link NativeObjectReference}.
+     */
+    private final NativeObjectReference resource;
+
+    static {
+        RocLibrary.loadLibrary();
+        thread.start();
+
+        /* add a ShutdownHook for closing all NativeObjects that are still open */
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            thread.closeAll();
+            try {
+                thread.join(MAX_JOIN_TIMEOUT_MS);
+            } catch (InterruptedException e) {}
+        }));
```

Comment:
Maybe better hide this thread creation/termination in AutoCloseThread.getInstance()? I think this code belongs to AutoCloseThread.

Reply:
Sure!
Anyway according to official Java documentation all registered shutdown hooks will run all concurrently and in unspecified order so there could be a problem if for example a user adds its own shutdown hook which makes use of roc objects inside it.
In my mind there could be three solutions for that:

1. Add a note on the README documenting explicitely to not use any roc object inside a user defined shutdown hook
2. Remove the addShutDownHook mechanism and expose a public method for closing all objects (I like this solution less than all the others because we are potentially giving this responsability to a lazy user which will not use any closing mechanism during program execution and closing objects only at the very end)
3. Remove the addShutDownHook mechanism at all

Reply:
Interesting! Hm, given that this PR is quite big already, I would suggest the third approach - remove the addShutDownHook mechanism at all. At least for now.

If the user calls close(), there is no leaks. If the user forgets to call close(), there is no strict guarantee. Bindings will collect dangling objects in some typical use cases, but the correct way to use bindings is still to call close().

I think this is a good approach to start with.

Reply:
removed shutdownHook in https://github.com/roc-streaming/roc-java/pull/40